torchao/prototype/hqq/core.py

```diff
@@ -0,0 +1,184 @@
+import torch
+import math
+from torch import Tensor, float16, float32
+from typing import Union
+
+
+# Shrinking operator (proximal operator for the lp norm)
+def shrink_lp_op(x: Tensor, beta: float, lp_norm: float) -> Tensor:
+    if lp_norm == 1:
+        return torch.sign(x) * torch.nn.functional.relu(torch.abs(x) - 1.0 / beta)
+    else:
+        return torch.sign(x) * torch.nn.functional.relu(
+            torch.abs(x) - (1.0 / beta) * torch.pow(torch.abs(x), lp_norm - 1)
+        )
+
+
+# Proximal solver || W - dequantize(quantize(W))||_p^p
+@torch.inference_mode()
+def optimize_weights_proximal_legacy(
+    tensor: Tensor,
+    scale: Tensor,
+    zero: Tensor,
+    min_max: list,
+    axis: int = 0,
+    dtype: Union[torch.dtype, None] = None,
+    device: Union[str, None] = None,
+    verbose: bool = False,
+    opt_params: dict = {
+        "lp_norm": 0.7,
+        "beta": 1e1,
+        "kappa": 1.01,
+        "iters": 20,
+        "early_stop": True,
+    },
+) -> tuple:
+    lp_norm, beta, kappa, iters, early_stop = (
+        opt_params["lp_norm"],
+        opt_params["beta"],
+        opt_params["kappa"],
+        opt_params["iters"],
+        opt_params["early_stop"],
+    )
+
+    device = tensor.device if (device is None) else torch.device(device)
+
+    if dtype is None:
+        dtype = float16 if (device.type == "cuda") else float32
+
+    W_f = tensor.to(dtype=dtype, device=device)
+    scale = scale.to(dtype=dtype, device=device)
+    zero = zero.to(dtype=dtype, device=device)
+
+    best_error = 1e4
+    for i in range(iters):
+        W_q = torch.round(W_f * scale + zero).clamp(min_max[0], min_max[1])
+        W_r = (W_q - zero) / scale
+        W_e = shrink_lp_op(W_f - W_r, beta, lp_norm)
+        zero = torch.mean(W_q - (W_f - W_e) * scale, axis=axis, keepdim=True)
+        beta *= kappa
+
+        current_error = float(torch.abs(W_f - W_r).mean())
+        if verbose:
+            print("Iter " + str(i + 1), " | Error: " + str(current_error))
+        if early_stop:
+            if current_error < best_error:
+                best_error = current_error
+            else:
+                break
+
+    scale = scale.to(tensor.device)
+    zero = zero.to(tensor.device)
+    del W_f, W_q, W_r, W_e
+    torch.cuda.empty_cache()
+
+    W_q = torch.round(tensor * scale + zero).clamp(min_max[0], min_max[1])
+    return W_q, scale, zero
+
+
+# Default: fast with early stopping
+optimize_weights_proximal = optimize_weights_proximal_legacy
+
+
+# Mainly used to check if the group-size is divisible by numel()
+def is_divisible(val1: int, val2: int) -> bool:
+    return int(val2 * math.ceil(val1 / val2)) == val1
+
+
+# Converts hqq format W_dequant = (W_q - zero)*scale into affinequantized format: (W_q - mid_point)*scale_ao + zero_ao
+def convert_to_affinequantized_format(W_q, scale, zero, nbits, shape):
```

Comment:
nit: to be more precise this is int4 tensor_core_tiled layout format (for tinygemm), and this probably only works for `nbits == 4`?

Reply:
I am not sure, I was just following the code in the torchao repo and noticed that it uses the mid-point version. 
In the hqq quantize function, you can choose if you want this format or the default format (`W_q - zero)*scale` via the `raw_output` argument. What format is used for lowers bits in ao?

Reply:
So I updated the example to run with mutliple nbits and it looks like it's working fine with the mid-point format across all nbits: https://github.com/pytorch/ao/pull/605/commits/e0226544b6e099e30cef2762ebe8bfd4293e36a0

Reply:
oh OK it should works for PlainLayoutType I think. it's an implementation detail in tinygemm though, where zero point domain is `ZeroPointDomain.FLOAT`

> What format is used for lowers bits in ao?

we haven't decided yet, but I think it will depend on benchmark results, for tinygemm the float zero_point domain (and preserve_zero=False) was chosen because it wants to use a fma instruction in cuda so it runs faster. and it's fine if we only care about linear quantization.

if we want to use quantization for conv, then we'd need to preserve zero: https://github.com/pytorch/ao/blob/5f35645bb75b67dd217c1cf17495ecb4a18b82fa/torchao/quantization/quant_primitives.py#L536 and set zero point domain to INT as well so that padding does not introduce loss

Reply:
I see, we can decide on this later The format would depend also on the cuda kernels for lower bits when they are available. Maybe we can follow the BitBlas approach, they have an argument `zeros_mode` which defines how the dequantization step is performed. They support different formats including `(Wq - zero)*scale` https://github.com/microsoft/BitBLAS/blob/main/docs/PythonAPI.md